src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1400,16 +1403,34 @@ private static class Key {
             File file;
             final boolean utf8;
 
-            public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
+            public Key(File file, BasicFileAttributes attrs, ZipCoder zc)
+                    throws IOException {
                 this.attrs = attrs;
-                this.file = file;
+                Object fk = attrs.fileKey();
+                this.file = (fk != null) ? file : getCanonicalFile(file);
                 this.utf8 = zc.isUTF8();
             }
 
+            @SuppressWarnings("removal")
+            private static File getCanonicalFile(File file) throws IOException {
+                if (System.getSecurityManager() == null) {
+                    return file.getCanonicalFile();
+                } else {
+                    try {
+                        return AccessController.doPrivileged(
+                                (PrivilegedExceptionAction<File>) file::getCanonicalFile);
+                    } catch (PrivilegedActionException pae) {
+                        throw new IOException(pae.getException());
+                    }
+                }
+            }
+
             public int hashCode() {
                 long t = utf8 ? 0 : Long.MAX_VALUE;
                 t += attrs.lastModifiedTime().toMillis();
```

Comment:
ZipFile cannot tolerate a zip file being changed while in use. I realise you don't want to change this part of the code but I think it is desperately needs a comment in both hashCode and equals to explain why the last modified time is included.